src/transformers/models/zamba/modeling_zamba.py

```diff
@@ -0,0 +1,1758 @@
+            query_states,
+            key_states,
+            value_states,
+            attn_mask=causal_mask,
+            dropout_p=self.attention_dropout if self.training else 0.0,
+            # The q_len > 1 is necessary to match with AttentionMaskConverter.to_causal_4d that does not create a causal mask in case q_len == 1.
+            is_causal=self.is_causal and attention_mask is None and q_len > 1,
+            scale=softmax_scale,
+        )
+
+        attn_output = attn_output.transpose(1, 2).contiguous()
+        attn_output = attn_output.view(bsz, q_len, 2 * self.hidden_size)
+
+        attn_output = self.o_proj(attn_output)
+
+        return attn_output, None, past_key_value
+
+
+ZAMBA_ATTENTION_CLASSES = {
+    "eager": ZambaAttention,
+    "flash_attention_2": ZambaFlashAttention2,
+    "sdpa": ZambaSdpaAttention,
+}
+
+
+class ZambaMambaMixer(nn.Module):
+    """
+    Compute ∆, A, B, C, and D the state space parameters and compute the `contextualized_states`.
+    A, D are input independent (see Mamba paper [1] Section 3.5.2 "Interpretation of A" for why A isn't selective)
+    ∆, B, C are input-dependent (this is a key difference between Mamba and the linear time invariant S4,
+    and is why Mamba is called **selective** state spaces)
+    """
+
+    def __init__(self, config: ZambaConfig, layer_idx):
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        self.hidden_size = config.hidden_size
+        self.ssm_state_size = config.mamba_d_state
+        self.conv_kernel_size = config.mamba_d_conv
+        self.intermediate_size = config.mamba_expand * config.hidden_size
+        self.time_step_rank = config.mamba_dt_rank
+        self.n_mamba_heads = config.n_mamba_heads
+        self.use_conv_bias = config.mamba_conv_bias
+        self.use_bias = config.mamba_proj_bias
+        self.conv1d = nn.Conv1d(
+            in_channels=self.intermediate_size,
+            out_channels=self.intermediate_size,
+            bias=self.use_conv_bias,
+            kernel_size=self.conv_kernel_size,
+            groups=self.intermediate_size,
+            padding=self.conv_kernel_size - 1,
+        )
+
+        self.activation = config.hidden_mamba_act
+        self.act = ACT2FN[config.hidden_mamba_act]
+
+        self.use_fast_kernels = config.use_mamba_kernels
+
+        # projection of the input hidden states
+        self.in_proj = nn.Linear(self.hidden_size, self.intermediate_size * 2, bias=self.use_bias)
+        # selective projection used to make dt, B and C input dependent
+        self.x_proj_weight = nn.Parameter(
+            (
+                torch.rand(
+                    self.n_mamba_heads,
+                    self.time_step_rank + self.ssm_state_size * 2,
+                    self.intermediate_size // self.n_mamba_heads,
+                )
+                - 0.5
+            )
+            * 2
+            / self.intermediate_size**0.5
+        )
```

Comment:
this should not be here. We initialize the weights in https://github.com/huggingface/transformers/blob/d2ea8f1afe70924fc2a5b07be14ca6c115a74213/src/transformers/models/mamba/modeling_mamba.py#L389

Reply:
should be alinear

Reply:
We used `nn.Parameter` to define weights as this allows to easily introduce an extra dimension associated to the mamba head (see description of mamba heads in one of the comments below). We originally used linears to define these weights and then re-shaped them to obtain the head dimension, but we found that this approach led to errors when training the model with DeepSpeed. Defining instead the parameters through `nn.Parameter` worked fine. We also noticed that `MambaMixer` uses `nn.Parameter` to define part of the weights (e.g. in this line https://github.com/Zyphra/transformers_zamba/blob/93e538ae2e307c2fb626fae4b914e25c58b47c0e/src/transformers/models/mamba/modeling_mamba.py#L110).

We are happy to define these weights starting from linear in a way that works with DeepSpeed but this will lead to different names of the weights (e.g. `x_proj.weight` instead of `x_proj_weight`) and, as mentioned in a comment above, this would involve replacing about 50 checkpoints stored on the HuggingFace hub. If the current implementation should still be modified, could you please let us know what we can do so that we preserve the weight names? We appreciate your help and effort with this.